peps/pep-0789.rst

```diff
@@ -0,0 +1,468 @@
+and so we never got a chance to create and raise an ``ExceptionGroup(...,
+[RuntimeError()])``.
+
+In a user-defined context manager
+---------------------------------
+
+Yielding across a CancelScope can be safe, if and only if you're using the
+generator to implement a context manager - in this case any propagating
+exceptions will be redirected to the expected task. [#redirected]_
+
+.. [#redirected] via e.g. ``contextlib.[async]contextmanager``,
+    or moral equivalents such as ``@pytest.fixture``
+
+We've also implemented a lint rule -- the amusingly named ``ASYNC101`` rule in
+`flake8-async <https://pypi.org/project/flake8-async/>`__ -- which warns against
+yielding across cancel scopes.  Could user education be sufficient to avoid
+these problems?  Unfortunately not: while it's safe to define context managers
+which yield across a cancel scope, that user-defined context manager now wraps a
+cancel scope, and is therefore unsafe to yield across!
+
+This regularly arises in practice, because 'run some background tasks for the
+duration of this context' is a very common pattern in structured concurrency.
+We saw that in ``combined_iterators()`` above; and have seen this bug in
+multiple implementations of the websocket protocol:
+
+.. code-block:: python
+
+    async def get_messages(websocket_url):
+        # The websocket protocol requires background tasks to manage the socket heartbeat
+        async with open_websocket(websocket_url) as ws:  # contains a TaskGroup!
+            while True:
+                yield await ws.get_message()
+
+    async with open_websocket(websocket_url) as ws:
+        async for message in get_messages(ws):
+            ...
+
+Restating the problem
+=====================
+
+Here's the fundamental issue: yield suspends a call frame. It only makes sense
+to yield in a leaf frame -- i.e., if your call stack goes like A -> B -> C, then
+you can suspend C, but you can't suspend B while leaving C running.
+
+But, TaskGroup is a kind of "concurrent call" primitive, where a single frame
+can have multiple child frames that run concurrently. This means that if we
+allow people to mix yield and TaskGroup, then we can end up in exactly this
+situation, where B gets suspended but C is actively running. This is
+nonsensical, and causes serious practical problems (e.g., if C raises an
+exception, we have no way to propagate it).
+
+This is a fundamental incompatibility between generator control flow and
+structured concurrency control flow, not something we can fix by tweaking our
+APIs. The only solution seems to be to forbid yield inside a TaskGroup.
+Although timeouts don't leave a child task running, the close analogy and
+related problems lead us to conclude that yield should be forbidden inside all
+CancelScopes, not only TaskGroups.
+
+Solution
+========
+
+We propose:
+
+1. a new context manager, ``with sys.block_yields(reason): ...`` which will
```

Comment:
Does this include both sync and async yield? If only async, maybe `block_async_yields()` is a better name. If the former, that puts a lot of restrictions on what you can do inside e.g. `with asyncio.timeout`.

Reply:
It includes `yield` in both sync and async generators, yes.  I'll add a note to clarify this.